Mathlib/Algebra/Lie/Weights.lean

```diff
@@ -249,6 +251,135 @@ instance [LieAlgebra.IsNilpotent R L] [IsNoetherian R M] :
+      let f₂ : Module.End R (L ⊗[R] M) := φ.lTensor L
+      replace hN : f₁ ^ N = 0 := by ext; simp [hN]
+      have h₁ : Commute f₁ f₂ := by ext; simp
+      have h₂ : φ ∘ₗ toModuleHom R L M = toModuleHom R L M ∘ₗ (f₁ + f₂) := by ext; simp
+      obtain ⟨q, hq⟩ := h₁.add_pow_dvd_pow_of_pow_eq_zero_right (N + k).le_succ hN
+      use toModuleHom R L M (q (y ⊗ₜ m))
+      change (φ ^ k).comp ((toModuleHom R L M : L ⊗[R] M →ₗ[R] M)) _ = _
+      simp [LinearMap.commute_pow_left_of_commute h₂, LinearMap.comp_apply (g := (f₁ + f₂) ^ k),
+        ← LinearMap.comp_apply (g := q), ← LinearMap.mul_eq_comp, ← hq] }
+
+variable {M} in
+lemma mem_posFittingCompOf [LieAlgebra.IsNilpotent R L] (x : L) (m : M) :
+    m ∈ posFittingCompOf R M x ↔ ∀ (k : ℕ), ∃ n, (toEndomorphism R L M x ^ k) n = m := by
+  simp [posFittingCompOf]
+
+@[simp] lemma posFittingCompOf_le_lowerCentralSeries [LieAlgebra.IsNilpotent R L] (x : L) (k : ℕ) :
+    posFittingCompOf R M x ≤ lowerCentralSeries R L M k := by
+  suffices : ∀ m l, (toEndomorphism R L M x ^ l) m ∈ lowerCentralSeries R L M l
+  · intro m hm
+    obtain ⟨n, rfl⟩ := (mem_posFittingCompOf R x m).mp hm k
+    exact this n k
+  intro m l
+  induction' l with l ih; simp
+  simp only [lowerCentralSeries_succ, pow_succ, LinearMap.mul_apply]
+  exact LieSubmodule.lie_mem_lie _ ⊤ (LieSubmodule.mem_top x) ih
+
+@[simp] lemma posFittingCompOf_eq_bot_of_isNilpotent
+    [LieAlgebra.IsNilpotent R L] [IsNilpotent R L M] (x : L) :
+    posFittingCompOf R M x = ⊥ := by
+  simp_rw [eq_bot_iff, ← iInf_lowerCentralSeries_eq_bot_of_isNilpotent, le_iInf_iff,
+    posFittingCompOf_le_lowerCentralSeries, forall_const]
+
+lemma exists_coe_posFittingCompOf_eq_of_isArtinian
+    [LieAlgebra.IsNilpotent R L] [IsArtinian R M] (x : L) :
+    ∃ (k : ℕ), posFittingCompOf R M x = LinearMap.range (toEndomorphism R L M x ^ k) :=
+  (toEndomorphism R L M x).exists_range_pow_eq_iInf
+
+variable (L)
+
+/-- If `M` is a representation of a nilpotent Lie algebra `L` with coefficients in `R`, then
+`posFittingComp R L M` is the span of the positive Fitting components of the action of `x` on `M`,
+as `x` ranges over `L`.
+
+It is a Lie submodule because `L` is nilpotent. -/
+def posFittingComp [LieAlgebra.IsNilpotent R L] : LieSubmodule R L M :=
+  ⨆ x, posFittingCompOf R M x
+
+lemma mem_posFittingComp [LieAlgebra.IsNilpotent R L] (m : M) :
+    m ∈ posFittingComp R L M ↔ m ∈ ⨆ (x : L), posFittingCompOf R M x := by
+  rfl
+
+lemma posFittingCompOf_le_posFittingComp [LieAlgebra.IsNilpotent R L] (x : L) :
+    posFittingCompOf R M x ≤ posFittingComp R L M := by
+  rw [posFittingComp]; exact le_iSup (posFittingCompOf R M) x
+
+lemma posFittingComp_le_iInf_lowerCentralSeries [LieAlgebra.IsNilpotent R L] :
+    posFittingComp R L M ≤ ⨅ k, lowerCentralSeries R L M k := by
+  simp [posFittingComp]
+
+/-- See also `LieModule.iSup_ucs_eq_weightSpace_zero`. -/
+@[simp] lemma iInf_lowerCentralSeries_eq_posFittingComp
```

Comment:
Is this original research, in this generality? Do you think you should somewhere be documenting what you are proving about this positive Fitting component? Maybe a section in the module docstring or something?